[@property] Set registered initial values to RenderStyle

anttijk opened the pull request:
#### ffe6d89e6fa50326715ba03ddd7134bc0d891dbc
<pre>
[@property] Set registered initial values to RenderStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=251575">https://bugs.webkit.org/show_bug.cgi?id=251575</a>
rdar://104953759

Reviewed by Antoine Quint.

Remove the need to provide the registry to RenderStyle::customPropertyValue().

This is done by maintaining a prototype style in CustomPropertyRegistry that is cloned
to make the initial styles in the style resolver.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-registered-custom-properties-expected.txt:

This also fixes some bugs.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendCustomProperty):
(WebCore::CSSPropertyAnimation::propertyRequiresBlendingForAccumulativeIteration):
(WebCore::CSSPropertyAnimation::propertiesEqual):
(WebCore::CSSPropertyAnimation::canPropertyBeInterpolated):
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::customPropertyValue):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::createPtrWithRegisteredInitialValues):
(WebCore::RenderStyle::inheritIgnoringCustomPropertiesFrom):
(WebCore::changedCustomPaintWatchedProperty):
(WebCore::RenderStyle::deduplicateCustomProperties):

Also dedupe non-inherited properties.

(WebCore::RenderStyle::setCustomPropertyValue):

Make this a single function.

(WebCore::RenderStyle::customPropertyValue const):
(WebCore::RenderStyle::deduplicateInheritedCustomProperties): Deleted.
(WebCore::RenderStyle::setInheritedCustomPropertyValue): Deleted.
(WebCore::RenderStyle::setNonInheritedCustomPropertyValue): Deleted.
(WebCore::RenderStyle::customPropertyValueWithoutResolvingInitial const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/StyleCustomPropertyData.h:
(WebCore::StyleCustomPropertyData::setCustomPropertyValue):
* Source/WebCore/style/CustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::CustomPropertyRegistry):
(WebCore::Style::CustomPropertyRegistry::registerFromAPI):
(WebCore::Style::CustomPropertyRegistry::registerFromStylesheet):
(WebCore::Style::CustomPropertyRegistry::clearRegisteredFromStylesheets):
(WebCore::Style::CustomPropertyRegistry::initialValuePrototypeStyle const):

Maintain the initial value prototype style.

(WebCore::Style::CustomPropertyRegistry::invalidate):
* Source/WebCore/style/CustomPropertyRegistry.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialCustomProperty):
(WebCore::Style::BuilderCustom::applyInheritCustomProperty):
(WebCore::Style::BuilderCustom::applyValueCustomProperty):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForElement):

Clone the initial style from the prototype.

(WebCore::Style::Resolver::styleForPseudoElement):
(WebCore::Style::Resolver::defaultStyleForElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Canonical link: <a href="https://commits.webkit.org/259807@main">https://commits.webkit.org/259807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e0a407019fff6ce2293ad2ca25c89f0ea4bc341

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106012 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15068 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38845 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115194 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109915 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16492 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/6249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98220 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111765 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95525 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/27168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8322 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28520 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8812 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/6249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14436 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48063 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6785 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10356 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2638,41 +2651,39 @@ void RenderStyle::setColumnStylesFromPaginationMode(const Pagination::Mode& pagi
-    // Alternatively we could just initialize the registered initial values to each RenderStyle.
-    auto* registered = registry.get(name);
-    return registered ? registered->initialValue.get() : nullptr;
+    if (isInherited)
+        set(m_rareInheritedData);
+    else
+        set(m_rareNonInheritedData);
```
I'd use a ternary operator here, not a big deal obviously:
```suggestion
    set(isInherited ? m_rareInheritedData : m_rareNonInheritedData);
```

anttijk replied:
That doesn't work since m_rareInheritedData and m_rareNonInheritedData have different types. The lambda here is a template really.

graouts commented on Source/WebCore/style/CustomPropertyRegistry.cpp:
```diff
@@ -71,8 +73,9 @@ bool CustomPropertyRegistry::registerFromAPI(CSSRegisteredCustomProperty&& prope
     if (success) {
-        m_scope.didChangeStyleSheetEnvironment();
+        invalidate();
         notifyAnimationsOfCustomPropertyRegistration(property.name);
```
Should we have `notifyAnimationsOfCustomPropertyRegistration()` called from `invalidate()`? I see you do this also in `registerFromStylesheet()` below.

anttijk replied:
Good point.

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-animation-expected.txt:
```diff
@@ -8,7 +8,7 @@ PASS Ongoing animation picks up redeclared inherits flag
 PASS Transitioning from initial value
 PASS Transitioning from specified value
-FAIL Transition triggered by initial value change assert_equals: expected "150px" but got "200px"
+PASS Transition triggered by initial value change
```
🎉

anttijk committed:
[@property] Set registered initial values to RenderStyle
https://bugs.webkit.org/show_bug.cgi?id=251575
rdar://104953759

Reviewed by Antoine Quint.

Remove the need to provide the registry to RenderStyle::customPropertyValue().

This is done by maintaining a prototype style in CustomPropertyRegistry that is cloned
to make the initial styles in the style resolver.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-registered-custom-properties-expected.txt:

This also fixes some bugs.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendCustomProperty):
(WebCore::CSSPropertyAnimation::propertyRequiresBlendingForAccumulativeIteration):
(WebCore::CSSPropertyAnimation::propertiesEqual):
(WebCore::CSSPropertyAnimation::canPropertyBeInterpolated):
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
(WebCore::CSSVariableReferenceValue::resolveVariableReference const):
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::customPropertyValue):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::createPtrWithRegisteredInitialValues):
(WebCore::RenderStyle::inheritIgnoringCustomPropertiesFrom):
(WebCore::changedCustomPaintWatchedProperty):
(WebCore::RenderStyle::deduplicateCustomProperties):

Also dedupe non-inherited properties.

(WebCore::RenderStyle::setCustomPropertyValue):

Make this a single function.

(WebCore::RenderStyle::customPropertyValue const):
(WebCore::RenderStyle::deduplicateInheritedCustomProperties): Deleted.
(WebCore::RenderStyle::setInheritedCustomPropertyValue): Deleted.
(WebCore::RenderStyle::setNonInheritedCustomPropertyValue): Deleted.
(WebCore::RenderStyle::customPropertyValueWithoutResolvingInitial const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/StyleCustomPropertyData.h:
(WebCore::StyleCustomPropertyData::setCustomPropertyValue):
* Source/WebCore/style/CustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::CustomPropertyRegistry):
(WebCore::Style::CustomPropertyRegistry::registerFromAPI):
(WebCore::Style::CustomPropertyRegistry::registerFromStylesheet):
(WebCore::Style::CustomPropertyRegistry::clearRegisteredFromStylesheets):
(WebCore::Style::CustomPropertyRegistry::initialValuePrototypeStyle const):

Maintain the initial value prototype style.

(WebCore::Style::CustomPropertyRegistry::invalidate):
* Source/WebCore/style/CustomPropertyRegistry.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialCustomProperty):
(WebCore::Style::BuilderCustom::applyInheritCustomProperty):
(WebCore::Style::BuilderCustom::applyValueCustomProperty):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::styleForElement):

Clone the initial style from the prototype.

(WebCore::Style::Resolver::styleForPseudoElement):
(WebCore::Style::Resolver::defaultStyleForElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Canonical link: https://commits.webkit.org/259807@main
